Make some float methods unstable `const fn`

eduardosm opened the pull request:
Some float methods are now `const fn` under the `const_float_methods` feature gate.

I also made some unstable methods `const fn`, keeping their constness under their respective feature gate.

In order to support `min`, `max`, `abs` and `copysign`, the implementation of some intrinsics had to be moved from Miri to rustc_const_eval (cc @RalfJung).

Tracking issue: https://github.com/rust-lang/rust/issues/130843

```rust
impl <float> {
    // #[feature(const_float_methods)]
    pub const fn recip(self) -> Self;
    pub const fn to_degrees(self) -> Self;
    pub const fn to_radians(self) -> Self;
    pub const fn max(self, other: Self) -> Self;
    pub const fn min(self, other: Self) -> Self;
    pub const fn clamp(self, min: Self, max: Self) -> Self;
    pub const fn abs(self) -> Self;
    pub const fn signum(self) -> Self;
    pub const fn copysign(self, sign: Self) -> Self;

    // #[feature(float_minimum_maximum)]
    pub const fn maximum(self, other: Self) -> Self;
    pub const fn minimum(self, other: Self) -> Self;

    // Only f16/f128 (f32/f64 already const)
    pub const fn is_sign_positive(self) -> bool;
    pub const fn is_sign_negative(self) -> bool;
    pub const fn next_up(self) -> Self;
    pub const fn next_down(self) -> Self;
}
```

r? libs-api

try-job: dist-s390x-linux

RalfJung commented on compiler/rustc_const_eval/src/interpret/intrinsics.rs:
```diff
@@ -440,6 +455,47 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
+                    _ => bug!(),
+                };
+                self.write_scalar(adjust_nan(self, res, &[a, b]), dest)?;
+            }
```
With just 2 types, the code duplication was bearable, but with 4 float types we really should have a single shared implementation that is generic over `rustc_apfloat::Float`. Can you try that?

RalfJung commented on compiler/rustc_const_eval/src/interpret/intrinsics.rs:
```diff
@@ -4,6 +4,7 @@
 
 use std::assert_matches::assert_matches;
 
+use rustc_apfloat::Float;
```
Please keep this qualified, it is only used a few times and it is useful to see immediately that this is an apfloat trait.

RalfJung commented:
r=me on the interpreter and Miri parts.

eduardosm commented:
I made `clamp` const too, using an approach similar to https://github.com/rust-lang/rust/pull/130611 to handle the assert formatting.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11020813709/job/30606476252) [(plain)](https://github.com/rust-lang/rust/commit/bf34851ab57fef3a72af86bbf3be923f8135f0b7/checks/30606476252/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:30ca74372d8b771363f68f939a58b017a592fae4f69398600dc51145997160f03e9652051f957840c41898984a88855e9757fa23464703a5a4ba21316ddebb04:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
failures:

---- [assembly] tests/assembly/x86_64-floating-point-clamp.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-18/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/assembly/x86_64-floating-point-clamp/x86_64-floating-point-clamp.s" "/checkout/tests/assembly/x86_64-floating-point-clamp.rs" "--check-prefix=CHECK" "--check-prefix" "NONMSVC" "--allow-unused-prefixes" "--dump-input-context" "100"
--- stderr -------------------------------
/checkout/tests/assembly/x86_64-floating-point-clamp.rs:22:17: error: CHECK-NEXT: expected string not found in input
/checkout/tests/assembly/x86_64-floating-point-clamp.rs:22:17: error: CHECK-NEXT: expected string not found in input
 // CHECK-NEXT: maxss xmm1, xmm0
                ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/assembly/x86_64-floating-point-clamp/x86_64-floating-point-clamp.s:45:12: note: scanning from here
 movss xmm1, dword ptr [rip + .LCPI1_1]
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/assembly/x86_64-floating-point-clamp/x86_64-floating-point-clamp.s:48:2: note: possible intended match here
 maxss xmm1, dword ptr [rsp + 4]

Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/assembly/x86_64-floating-point-clamp/x86_64-floating-point-clamp.s
Check file: /checkout/tests/assembly/x86_64-floating-point-clamp.rs


-dump-input=help explains the following input dump.
Input was:
<<<<<<
           1:  .text 
           1:  .text 
           2:  .intel_syntax noprefix 
           3:  .file "x86_64_floating_point_clamp.1f7dbb53b27604f7-cgu.0" 
           4:  .section .text.clamp_demo,"ax",@progbits 
           5:  .globl clamp_demo 
           6:  .p2align 4, 0x90 
           7:  .type clamp_demo,@function 
           8: clamp_demo: 
           9:  .cfi_startproc 
          10:  sub rsp, 24 
          11:  .cfi_def_cfa_offset 32 
          12:  movss dword ptr [rsp + 16], xmm2 
          13:  movss dword ptr [rsp + 12], xmm1 
          14:  movss dword ptr [rsp + 20], xmm0 
          15:  movaps xmm0, xmm1 
          16:  movaps xmm1, xmm2 
          17:  call qword ptr [rip + _RNvNvMNtCsGDmjXnHhwR_4core3f32f5clamp12assert_at_rt@GOTPCREL] 
          18:  movss xmm1, dword ptr [rsp + 12] 
          19:  maxss xmm1, dword ptr [rsp + 20] 
          20:  movss xmm0, dword ptr [rsp + 16] 
          21:  minss xmm0, xmm1 
          22:  add rsp, 24 
          23:  .cfi_def_cfa_offset 8 
          24:  ret 
          25: .Lfunc_end0: 
          26:  .size clamp_demo, .Lfunc_end0-clamp_demo 
          27:  .cfi_endproc 
          28:  
          29:  .section .rodata.cst4,"aM",@progbits,4 
          30:  .p2align 2, 0x0 
          31: .LCPI1_0: 
          32:  .long 0x3f800000 
          33: .LCPI1_1: 
          34:  .long 0x40000000 
          35:  .section .text.clamp12_demo,"ax",@progbits 
          36:  .globl clamp12_demo 
          37:  .p2align 4, 0x90 
          38:  .type clamp12_demo,@function 
          39: clamp12_demo: 
          40:  .cfi_startproc 
          41:  push rax 
          42:  .cfi_def_cfa_offset 16 
          43:  movss dword ptr [rsp + 4], xmm0 
          44:  movss xmm0, dword ptr [rip + .LCPI1_0] 
          45:  movss xmm1, dword ptr [rip + .LCPI1_1] 
next:22'0                X~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
          46:  call qword ptr [rip + _RNvNvMNtCsGDmjXnHhwR_4core3f32f5clamp12assert_at_rt@GOTPCREL] 
next:22'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          47:  movss xmm1, dword ptr [rip + .LCPI1_0] 
next:22'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          48:  maxss xmm1, dword ptr [rsp + 4] 
next:22'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
next:22'1      ?                                possible intended match
          49:  movss xmm0, dword ptr [rip + .LCPI1_1] 
next:22'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          50:  minss xmm0, xmm1 
next:22'0     ~~~~~~~~~~~~~~~~~~
          51:  pop rax 
next:22'0     ~~~~~~~~~
          52:  .cfi_def_cfa_offset 8 
next:22'0     ~~~~~~~~~~~~~~~~~~~~~~~
          53:  ret 
next:22'0     ~~~~~
          54: .Lfunc_end1: 
next:22'0     ~~~~~~~~~~~~~
          55:  .size clamp12_demo, .Lfunc_end1-clamp12_demo 
next:22'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          56:  .cfi_endproc 
next:22'0     ~~~~~~~~~~~~~~
next:22'0     ~
next:22'0     ~
          58:  .ident "rustc version 1.83.0-nightly (50cf3b5e9 2024-09-24)" 
next:22'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          59:  .section ".note.GNU-stack","",@progbits 
next:22'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
------------------------------------------



```

</details>

tgross35 commented:
> If this is accepted, a tracking issue should be created (for now I set `issue = "none"`).

You can go ahead and do this in advance, I don't think there is any reason these would be rejected

tgross35 commented on library/core/src/num/f128.rs:
```diff
@@ -471,7 +471,8 @@ impl f128 {
     #[must_use]
     #[unstable(feature = "f128", issue = "116909")]
-    pub fn is_sign_positive(self) -> bool {
+    #[rustc_const_unstable(feature = "f128", issue = "116909")]
+    pub const fn is_sign_positive(self) -> bool {
```
Thanks for including f16 and f128 :) would you mind adding a commented out gate on methods from this feature? E.g.

```rust
    // #[rustc_const_unstable(feature = "const_float_methods", issue = "...")]
    #[rustc_const_unstable(feature = "f128", issue = "116909")]
    pub const fn is_sign_positive(self) -> bool {
```

We just do that to keep things somewhat linked if API falls under >1 gate.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11037921943/job/30659964477) [(plain)](https://github.com/rust-lang/rust/commit/c06c2551b9394e0e937e3a667361573c3525b4af/checks/30659964477/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 2.798 Building wheels for collected packages: reuse
#16 2.799   Building wheel for reuse (pyproject.toml): started
#16 3.047   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 3.048   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#16 3.049   Stored in directory: /tmp/pip-ephem-wheel-cache-oig7qr3b/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#16 3.051 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#16 3.443 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#16 3.444 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#16 DONE 3.5s
---
...
645  | /     /// Converts radians to degrees.
646  | |     ///
647  | |     /// ```
648  | |     /// #![feature(f128)]
657  | |     /// ```
658  | |     #[inline]
     | |_____________- other attributes here
     | |_____________- other attributes here
659  |       #[unstable(feature = "f128", issue = "116909")]
...
1316 |   }
     |   - the item list ends here

---

error[E0599]: no method named `to_bits` found for type `f128` in the current scope
   --> core/src/num/f128.rs:508:15
    |
508 |         (self.to_bits() & (1 << 127)) != 0
    |               ^^^^^^^ method not found in `f128`
error[E0599]: no method named `to_bits` found for type `f128` in the current scope
   --> core/src/num/f128.rs:551:25
    |
551 |         let bits = self.to_bits();
551 |         let bits = self.to_bits();
    |                         ^^^^^^^ method not found in `f128`

error[E0599]: no method named `to_bits` found for type `f128` in the current scope
   --> core/src/num/f128.rs:552:52
    |
552 |         if self.is_nan() || bits == Self::INFINITY.to_bits() {
    |                                                    ^^^^^^^ method not found in `f128`
error[E0599]: no function or associated item named `from_bits` found for type `f128` in the current scope
   --> core/src/num/f128.rs:564:15
    |
564 |         Self::from_bits(next_bits)
---

error[E0599]: no method named `to_bits` found for type `f128` in the current scope
   --> core/src/num/f128.rs:608:56
    |
608 |         if self.is_nan() || bits == Self::NEG_INFINITY.to_bits() {
    |                                                        ^^^^^^^ method not found in `f128`
error[E0599]: no function or associated item named `from_bits` found for type `f128` in the current scope
   --> core/src/num/f128.rs:620:15
    |
620 |         Self::from_bits(next_bits)
```

</details>

eduardosm commented:
Created tracking issue: https://github.com/rust-lang/rust/issues/130843

For f16/f32 methods, I placed commented gates under the same feature as f32/f64 (e.g., `is_sign_positive` under `const_float_classify`, `next_up` under `float_next_up_down`).

bors commented:
:umbrella: The latest upstream changes (presumably #130157) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

eduardosm commented:
I believe `is_sign_positive` / `is_sign_negative` do not need the second unstable gate since `const_float_classify` is now stable.

RalfJung commented on src/tools/miri/src/intrinsics/mod.rs:
```diff
@@ -249,28 +249,18 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
             }
 
-            "minnumf32" | "maxnumf32" | "copysignf32" => {
+            "copysignf32" => {
```
Btw, is there a reason you are not including `copy_sign` in this?

eduardosm replied:
I could move it to rustc_const_eval. But, do we want to make `copysign` and `abs` const? Ideally, we'd want to, but since they are implemented in libstd instead of libcore, I'm not sure what's the deal with those.

RalfJung replied:
I don't see why we wouldn't make them const.

I also don't know why they are in `std`, maybe the intrinsic is sometimes compiled to a libcall or so... let's just leave them where they are. :shrug:

eduardosm replied:
I'll make them const. But I find it quite weird that we have operations like min/max or `%` in libcore, while we are afraid to have operations like `abs` or `copysign` that can be trivially implemented with bitwise operations.

RalfJung replied:
I fully agree. Maybe file an issue for that?

RalfJung commented on compiler/rustc_const_eval/src/interpret/intrinsics.rs:
```diff
@@ -697,4 +747,54 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
+    }
+}
+
+enum FloatBinOp {
```
```suggestion
enum FloatBinIntrinsic {
```
We use `BinOp` for the MIR binops, IMO we shouldn't mix up that terminology.

Also please move this above the functions that use it.

RalfJung commented on tests/ui/consts/const-eval/float_methods.rs:
```diff
@@ -0,0 +1,46 @@
+//@ run-pass
```
```suggestion
//@ run-pass
//! Tests the float intrinsics: min, max, abs, copysign
```

RalfJung commented:
LGTM from the interpreter side.

r? libs-api

eduardosm commented:
Rebased to pick https://github.com/rust-lang/rust/pull/131256 and remove `#![feature(f16_const)]` and `#![feature(f128_const)]`.

bors commented:
:umbrella: The latest upstream changes (presumably #131581) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

tgross35 commented:
If libs-api can just okay adding `const` to the signatures, I'm happy to review the rest.

@rustbot label +I-libs-api-nominated

tgross35 commented on compiler/rustc_const_eval/src/interpret/intrinsics.rs:
```diff
@@ -438,6 +444,56 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
             }
 
+            sym::minnumf16 | sym::maxnumf16 | sym::copysignf16 => {
+                let op = match intrinsic_name {
+                    sym::minnumf16 => FloatBinIntrinsic::Min,
+                    sym::maxnumf16 => FloatBinIntrinsic::Max,
+                    sym::copysignf16 => FloatBinIntrinsic::Copysign,
+                    _ => bug!(),
+                };
+                self.float_bin_op_intrinsic::<rustc_apfloat::ieee::Half>(op, args, dest)?;
```
Nit: If you import `rustc_apfloat::ieee::{Half, Single, Double, Quad}` then is it's probably just as compact to have flat match branches, rather than nesting a `match` and having the symbol names repeated twice.

```rust
            sym::minnumf16 => self.float_bin_op_intrinsic::<Half>(FloatBinIntrinsic::Min, args, dest)?,
            sym::minnumf32 => self.float_bin_op_intrinsic::<Single>(FloatBinIntrinsic::Min, args, dest)?,
            sym::minnumf64 => self.float_bin_op_intrinsic::<Double>(FloatBinIntrinsic::Min, args, dest)?,
            sym::minnumf128=> self.float_bin_op_intrinsic::<Quad>(FloatBinIntrinsic::Min, args, dest)?,
            // ...
```

eduardosm replied:
`fmt` won't keep them single line :(

```rust
sym::minnumf16 => {
    self.float_bin_op_intrinsic::<Half>(FloatBinIntrinsic::Min, args, dest)?
}
sym::minnumf32 => {
    self.float_bin_op_intrinsic::<Single>(FloatBinIntrinsic::Min, args, dest)?
}
sym::minnumf64 => {
    self.float_bin_op_intrinsic::<Double>(FloatBinIntrinsic::Min, args, dest)?
}
sym::minnumf128 => {
    self.float_bin_op_intrinsic::<Quad>(FloatBinIntrinsic::Min, args, dest)?
}
```

Another option would be to use a dedicated method for each operation, which would avoid the lengthy `FloatBinIntrinsic::...`.

```rust
sym::minnumf16 => self.float_min_intrinsic::<Half>(args, dest)?,
sym::minnumf32 => self.float_min_intrinsic::<Single>(args, dest)?,
sym::minnumf64 => self.float_min_intrinsic::<Double>(args, dest)?,
sym::minnumf128 => self.float_min_intrinsic::<Quad>(args, dest)?,
```

tgross35 replied:
I think that looks cleaner, nice.

BurntSushi commented:
The unstable const APIs LGTM (on the basis that we're doing the same for `f32` and `f64`).

tgross35 commented:
Thanks @BurntSushi for confirming.

@eduardosm I think something changed recently where you don't need `rustc_const_unstable` if the function is already `unstable` and the feature should be the same. I think this is the case for `float_minimum_maximum`, `f16` and `f128` - would you be able to test removing the `const_unstable` gates?

Either way, just needs a squash then lgtm.

RalfJung commented on compiler/rustc_const_eval/src/interpret/intrinsics.rs:
```diff
@@ -697,4 +718,73 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
+}
+
+// Performs appropriate non-deterministic adjustments of NaN results.
+fn adjust_nan<'tcx, M, F1, F2>(this: &InterpCx<'tcx, M>, f: F2, inputs: &[F1]) -> F2
+where
+    M: Machine<'tcx>,
+    F1: rustc_apfloat::Float + rustc_apfloat::FloatConvert<F2>,
```
We already have this exact same function [in Miri](https://github.com/rust-lang/rust/blob/c4ce8c114b06840c3521a189ee44958b713fb33a/src/tools/miri/src/operator.rs#L119-L121). :joy: 

Can you instead make this a `pub fn` on the `InterpCx` type, and adjust the Miri code to also use it?

RalfJung replied:
Probably `operator.rs` is the best place for this, and there are some `M::generate_nan` calls that could also use this.

RalfJung commented on library/core/src/num/f128.rs:
```diff
@@ -627,8 +633,10 @@ impl f128 {
     #[inline]
     #[unstable(feature = "f128", issue = "116909")]
+    // #[rustc_const_unstable(feature = "const_float_methods", issue = "130843")]
+    #[rustc_const_unstable(feature = "f128", issue = "116909")]
```
It probably makes sense to use `const_float_methods` for `rustc_const_unstable` here, that way one has to enable both feature gates to const-call this function.

eduardosm committed:
De-duplicate and move `adjust_nan` to `InterpCx`

eduardosm committed:
Make some float methods unstable `const fn`

Some float methods are now `const fn` under the `const_float_methods` feature gate.

In order to support `min`, `max`, `abs` and `copysign`, the implementation of some intrinsics had to be moved from Miri to rustc_const_eval.

RalfJung commented:
I have pushed some further dedup to your branch, I hope you don't mind. Feel free to squash this into your first commit or keep it separate, whatever you prefer.

eduardosm commented:
> I have pushed some further dedup to your branch, I hope you don't mind. Feel free to squash this into your first commit or keep it separate, whatever you prefer.

Thanks. My search for "fn adjust_nan" missed that "let adjust_nan" :)

> r=me for the library/ changes, with the note that I think we don't need `#[rustc_const_unstable(feature = "f16", issue = "116909")]` anymore on functions that are `#[unstable(feature = "f16", issue = "116909")]` (same for any other unstable functions where the stability and const-stability gates are the same).

I think this time I took care of every remaining redundant `rustc_const_unstable` (for methods touched in this PR).

RalfJung commented:
Thanks!
@bors r=RalfJung,tgross35 rollup

bors commented:
:pushpin: Commit 029a881c08bba829a13eb9bd91f450cda2df827d has been approved by `RalfJung,tgross35`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=RalfJung,tgross35 029a881c08bba829a13eb9bd91f450cda2df827d -->
<!-- homu: {"type":"Approved","sha":"029a881c08bba829a13eb9bd91f450cda2df827d","approver":"RalfJung,tgross35","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

jieyouxu commented:
I think this failed in the rollup https://github.com/rust-lang/rust/pull/131711#issuecomment-2412633662

jieyouxu commented:
@bors r-

jieyouxu commented:
@bors try

bors commented:
:hourglass: Trying commit 029a881c08bba829a13eb9bd91f450cda2df827d with merge b181e49445cdf73262cd4891a5f76986ef702bc6...
<!-- homu: {"type":"TryBuildStarted","head_sha":"029a881c08bba829a13eb9bd91f450cda2df827d","merge_sha":"b181e49445cdf73262cd4891a5f76986ef702bc6"} -->

compiler-errors commented:
@bors rollup=never

tgross35 commented:
Ah yeah yuck, every function in stdlib that takes a f16 or f128 in the signature needs `#[inline]` so LLVM doesn't try to codegen and crash on platforms where that is buggy. This includes the `assert_at_const` and `assert_at_rt` functions.

rust-log-analyzer commented:
The job **`dist-s390x-linux`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/11338594437/job/31532070921) [(plain)](https://github.com/rust-lang-ci/rust/commit/b181e49445cdf73262cd4891a5f76986ef702bc6/checks/31532070921/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config remote.upstream.fetch=+refs/heads/*:refs/remotes/upstream/*
file:.git/config submodule.library/backtrace.active=true
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
---
[RUSTC-TIMING] rustc_std_workspace_alloc test:false 0.028
   Compiling gimli v0.29.0
[RUSTC-TIMING] panic_abort test:false 0.046
   Compiling miniz_oxide v0.7.4
rustc-LLVM ERROR: Cannot select: 0x7f2d1458a0e0: f32 = fp16_to_fp 0x7f2d14589e40
  0x7f2d14589e40: i32 = fp_to_fp16 0x7f2d145894a0
    0x7f2d145894a0: f32,ch = CopyFromReg 0x7f2d25c82168, Register:f32 %1
      0x7f2d14589ba0: f32 = Register %1
In function: _ZN4core3f1621_$LT$impl$u20$f16$GT$5clamp15assert_at_const17hb8416e17aae8c18dE
error: could not compile `core` (lib)
warning: build failed, waiting for other jobs to finish...
[RUSTC-TIMING] alloc test:false 4.995
[RUSTC-TIMING] miniz_oxide test:false 1.099
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/11338594437/job/31532769891)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/11338594437/job/31532769891","builder_name":"checks-actions"} -->

RalfJung commented:
@bors r=RalfJung,tgross35 rollup

bors commented:
:pushpin: Commit c09ed3e767a73d83673790f74c357432fa44d320 has been approved by `RalfJung,tgross35`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=RalfJung,tgross35 c09ed3e767a73d83673790f74c357432fa44d320 -->
<!-- homu: {"type":"Approved","sha":"c09ed3e767a73d83673790f74c357432fa44d320","approver":"RalfJung,tgross35","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

RalfJung commented:
> Ah yeah yuck, every function in stdlib that takes a f16 or f128 in the signature still needs `#[inline]` so LLVM doesn't try to codegen and crash on platforms where that is buggy. This includes the `assert_at_const` and `assert_at_rt` functions.

Wow f16 / f128 are more unstable than I thought.^^ Is there an issue tracking specifically the codegen crashes? Can we have a central workaround instead of dozens of `#[inline]` everywhere where we will surely forget to remove some?

tgross35 commented:
> Wow f16 / f128 are more unstable than I thought.^^ Is there an issue tracking specifically the codegen crashes? Can we have a central workaround instead of dozens of `#[inline]` everywhere where we will surely forget to remove some?

The list of platforms where just seeing the types is enough to make LLVM crash is fortunately pretty limited, wasm and arm64ec being some notably more popular exceptions https://github.com/rust-lang/compiler-builtins/blob/cb060052ab7e4bad408c85d44be7e60096e93e38/configure.rs#L53-L77. AIUI however, we need the inline annotations anyway since Cranelift doesn't support the types, and having anything not inline would mean it can't compile std.

I'm not sure what would be better here.

RalfJung commented:
> I'm not sure what would be better here.

We can automatically mark functions as `inline` if they have a `f16`/`f128` argument. :see_no_evil:

tgross35 commented:
That works 😄

RalfJung commented:
Specifically, the logic [here](https://github.com/rust-lang/rust/blob/4beb1cf9e53102ad32793127316f0045a9c004f3/compiler/rustc_mir_transform/src/cross_crate_inline.rs#L17), which already makes "small" functions auto-`#[inline]`, could also check the signature.
